File: leetcode/Algorithms/797.AllPathsFromSourceToTarget/Solution.java
import java.util.*;

class Solution {
    public List<List<Integer>> allPathsSourceTarget(int[][] graph) {
        List<List<Integer>> result = new ArrayList<>();
        Queue<List<Integer>> queue = new LinkedList<>();
        queue.add(Collections.singletonList(0));
        while (queue.size() > 0) {
            List<Integer> tmp = queue.poll();
            if (tmp.get(tmp.size() - 1) == graph.length - 1) {
                result.add(tmp);
            } else {
                for (int n : graph[tmp.get(tmp.size() - 1)]) {
                    List<Integer> next = new ArrayList<>(tmp);
                    next.add(n);
                    queue.add(next);
                }
            }
        }
        return result;
    }
}

File: leetcode/Algorithms/797.AllPathsFromSourceToTarget/Solution.go
func allPathsSourceTarget(graph [][]int) [][]int {
	result := make([][]int, 0)
	dfs(graph, 0, make([]int, 1), &result)
	return result
}

func dfs(graph [][]int, index int, tmp []int, result *[][]int) {
	for _, n := range graph[index] {
		next := make([]int, len(tmp))
		copy(next, tmp)
		next = append(next, n)
		if n == len(graph)-1 {
			*result = append(*result, next)
		} else {
			dfs(graph, n, next, result)
		}
	}
}

